sympy/matrices/random.py

```diff
@@ -0,0 +1,2525 @@
+from random import Random
+
+from ..core import I as _i
+from ..core.mul import Mul as _multiply
+from ..functions import sqrt as _sqrt, re as _re, im as _im, \
+    transpose as _t, adjoint as _c
+from ..tensor import shape as _shape
+from .dense import eye as _eye
+from .inverse import _inv
+
+__all__ = 'projection', 'jordan', 'transposition', \
+          'permutation', 'elementary', 'rotation', 'reflection', \
+          'diagonal_normal', 'jordan_normal', 'isometry_normal', \
+          'triangular', 'invertible', 'singular', \
+          'idempotent', 'nilpotent', 'diagonalizable', 'trigonalizable', \
+          'orthogonal', 'unitary', 'normal', \
+          'symmetric', 'hermite', 'square', \
+          'regular_to_singular', 'complex_to_real', 'rand'
+
+
+_ALT = False
+_EPS = 1e-15
+
+
+# === random number generator functions ===
+
+
+rand = Random()
+r""" default random number generator
+
+    Explanation
+    ===========
+
+    the module default random generator
+    which is a instance of ``random.Random()``.
+    Hence, the state can be set by invoking ``rand.seed()``.
+
+
+    Examples
+    ========
+
+    >>> from sympy.matrices.random import rand, orthogonal
+
+    >>> rand.seed(1)
+    >>> orthogonal(3)
+    Matrix([
+    [-1,          0,          0],
+    [ 0,  sqrt(2)/2, -sqrt(2)/2],
+    [ 0, -sqrt(2)/2, -sqrt(2)/2]])
+
+    See Also
+    ========
+    projection
+
+"""
+
+
+def _rand(seed=None):
+    if hasattr(seed, 'sample'):
+        return seed
+    return rand if seed is None else Random(seed)
+
+
+def _sample(scalars, k=None, rng=None):
+    _k = k or 1
+    if hasattr(scalars, 'sample'):
+        smpl = scalars.sample(_k)
+    elif hasattr(rng, 'sample'):
+        smpl = rng.sample(scalars, _k)
+    else:
+        smpl = rand.sample(scalars, _k)
+    return smpl if k else smpl[0]
+
+
+# === helper ===
+
+
+def _set_value(o, i, j, v):
+    o[i, j] = v
+
+
+def _cs(scalar, real=True, rng=None):
+    if isinstance(scalar, (tuple, list)) and len(scalar) == 2:
+        c, s = scalar
+    elif _is_complex(scalar):
+        c, s = _re(scalar), _im(scalar)
+    else:
+        rng = _rand(rng)
+        c, s = scalar, _sample((-1, 1), rng=rng) * _sqrt(1 - scalar ** 2)
+    abs_cs = abs(c ** 2 + s ** 2 - 1)
+    if isinstance(scalar, (int, float, complex)):
+        if abs(scalar) > 1 or abs_cs > _EPS:
+            msg = "isometry scalar argument must have norm equal to 1"
+            msg += " or - if real - norm less than 1"
+            msg += " not abs%s=%s" % (str(scalar), str(abs_cs))
+            raise ValueError(msg)
+    return (c, s) if real else (c + s * _i,)
+
+
+def _is_complex(z):
+    return isinstance(z, complex) or \
+           (getattr(z, 'is_complex', False) is True and
+            getattr(z, 'is_real', True) is False)
```

Comment:
The `is_complex` and `is_real` attributes are part of the assumptions system. I'm not sure exactly what the intended purpose of this check is but `is_real` might not be what is wanted here.

Usually in SymPy it isn't necessary to try to handle built in types at the same time as SymPy types because wen just use sympify to turn everything into SymPy types e.g.:
```python
In [1]: from sympy.core.sympify import _sympify

In [2]: _sympify(1j)
Out[2]: 1.0⋅ⅈ

In [3]: _sympify(1j).is_real
Out[3]: False
```

Reply:
Good hint, thanks.